File: compiler/src/lexer/lexer.test.js
const Lexer = require('./lexer');
const TokenType = require('./libs/token-types');

describe('Lexer', () => {
    it('tokenizes identifiers and keywords', () => {
        const lexer = new Lexer('let x = 42;');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toEqual([
            TokenType.KEYWORD, // let
            TokenType.IDENTIFIER, // x
            TokenType.ASSIGNMENT, // =
            TokenType.NUMBER, // 42
            TokenType.SEMICOLON, // ;
            TokenType.EOF
        ]);
    });

    it('tokenizes template literals', () => {
        const lexer = new Lexer('const s = `hello ${name}`;');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_STRING)).toBe(true);
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_EXPRESSION)).toBe(true);
    });

    it('tokenizes optional chaining and nullish coalescing', () => {
        const lexer = new Lexer('a?.b ?? c;');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toContain(TokenType.OPTIONAL_CHAINING);
        expect(tokens.map(t => t.type)).toContain(TokenType.NULLISH_COALESCING);
    });

    it('tokenizes union and intersection types', () => {
        const lexer = new Lexer('let x: string | number & boolean;');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toContain(TokenType.UNION);
        expect(tokens.map(t => t.type)).toContain(TokenType.INTERSECTION);
    });

    it('throws on unterminated string', () => {
        const lexer = new Lexer('let s = "unterminated;');
        expect(() => lexer.tokenize()).toThrow(/Unterminated string/);
    });

    it('throws on unterminated template string', () => {
        const lexer = new Lexer('let s = `unterminated;');
        expect(() => lexer.tokenize()).toThrow(/Unterminated template string/);
    });
});

describe('Lexer - comprehensive coverage', () => {
    it('tokenizes all bracket types', () => {
        const lexer = new Lexer('() {} [] <>');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toEqual([
            TokenType.LEFT_PAREN, TokenType.RIGHT_PAREN,
            TokenType.LEFT_BRACE, TokenType.RIGHT_BRACE,
            TokenType.LEFT_BRACKET, TokenType.RIGHT_BRACKET,
            TokenType.LEFT_ANGLE, TokenType.RIGHT_ANGLE,
            TokenType.EOF
        ]);
    });

    it('tokenizes all punctuation', () => {
        const lexer = new Lexer('; , . : ?');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toEqual([
            TokenType.SEMICOLON, TokenType.COMMA, TokenType.DOT, TokenType.COLON, TokenType.QUESTION_MARK, TokenType.EOF
        ]);
    });

    it('tokenizes all basic operators', () => {
        const lexer = new Lexer('+ - * / % ! = == === != !== || && | &');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toEqual([
            TokenType.OPERATOR, TokenType.OPERATOR, TokenType.OPERATOR, TokenType.OPERATOR, TokenType.OPERATOR, TokenType.OPERATOR,
            TokenType.ASSIGNMENT, TokenType.OPERATOR, TokenType.OPERATOR, TokenType.OPERATOR, TokenType.OPERATOR,
            TokenType.OPERATOR, TokenType.OPERATOR, TokenType.UNION, TokenType.INTERSECTION, TokenType.EOF
        ]);
    });

    it('tokenizes all literal types', () => {
        const lexer = new Lexer('true false null undefined NaN Infinity');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toEqual([
            TokenType.KEYWORD, TokenType.KEYWORD, TokenType.KEYWORD, TokenType.KEYWORD, TokenType.KEYWORD, TokenType.KEYWORD, TokenType.EOF
        ]);
    });

    it('tokenizes numbers (integers and floats)', () => {
        const lexer = new Lexer('123 45.67 0.89 .5 6.');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toEqual([
            TokenType.NUMBER, TokenType.NUMBER, TokenType.NUMBER, TokenType.NUMBER, TokenType.NUMBER, TokenType.EOF
        ]);
    });

    it('tokenizes all keyword categories', () => {
        const code = [
            'let', 'if', 'match', 'immutable', 'requires', 'try', 'doc', 'string', 'import', 'this', 'in', 'with'
        ].join(' ');
        const lexer = new Lexer(code);
        const tokens = lexer.tokenize();
        // All should be KEYWORD except 'this' (object), 'in' (operator), 'with' (misc), which are also KEYWORD in your setup
        expect(tokens.every(t => t.type === TokenType.KEYWORD || t.type === TokenType.EOF)).toBe(true);
    });

    it('skips single-line comments', () => {
        const lexer = new Lexer('let x = 1; // this is a comment\nconst y = 2;');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.value === '//')).toBe(false);
        expect(tokens.map(t => t.type)).toContain(TokenType.KEYWORD);
        expect(tokens.map(t => t.type)).toContain(TokenType.IDENTIFIER);
    });

    it('skips multi-line comments', () => {
        const lexer = new Lexer('let x = 1; /* multi\nline\ncomment */ const y = 2;');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.value === '/*')).toBe(false);
        expect(tokens.map(t => t.type)).toContain(TokenType.KEYWORD);
        expect(tokens.map(t => t.type)).toContain(TokenType.IDENTIFIER);
    });

    it('throws on invalid character', () => {
        const lexer = new Lexer('let x = 1 @');
        expect(() => lexer.tokenize()).toThrow(/Invalid character/);
    });

    it('tokenizes consecutive operators and edge cases', () => {
        const lexer = new Lexer('a === b !== c == d != e');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toEqual([
            TokenType.IDENTIFIER, TokenType.OPERATOR, TokenType.IDENTIFIER, TokenType.OPERATOR, TokenType.IDENTIFIER,
            TokenType.OPERATOR, TokenType.IDENTIFIER, TokenType.OPERATOR, TokenType.IDENTIFIER, TokenType.EOF
        ]);
    });

    it('tokenizes type assertions and generics', () => {
        const lexer = new Lexer('x as string generic infer');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toContain(TokenType.KEYWORD);
        expect(tokens.map(t => t.value)).toContain('as');
        expect(tokens.map(t => t.value)).toContain('generic');
        expect(tokens.map(t => t.value)).toContain('infer');
    });
});

describe('Lexer - edge and missing cases', () => {
    it('tokenizes escaped quotes in strings', () => {
        const lexer = new Lexer("\"foo\\\"bar\" 'foo\\'bar'");
        const tokens = lexer.tokenize();
        expect(tokens.filter(t => t.type === TokenType.STRING).length).toBe(2);
    });

    it('tokenizes unicode escapes in strings', () => {
        const lexer = new Lexer('"\\u1234"');
        const tokens = lexer.tokenize();
        expect(tokens[0].type).toBe(TokenType.STRING);
    });

    it('handles nested template expressions', () => {
        const lexer = new Lexer('`a${b + `c${d}`}`');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_STRING)).toBe(true);
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_EXPRESSION)).toBe(true);
    });

    it('handles escaped backticks in template strings', () => {
        const lexer = new Lexer('`foo\\`bar`');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_STRING)).toBe(true);
    });

    it('tokenizes numbers in exponential notation', () => {
        const lexer = new Lexer('1e10 2.5e-3');
        const tokens = lexer.tokenize();
        // console.log(tokens);
        expect(tokens.filter(t => t.type === TokenType.NUMBER).length).toBe(2);
    });

    it('tokenizes hexadecimal, binary, and octal numbers', () => {
        const lexer = new Lexer('0x1A 0b101 0o77');
        const tokens = lexer.tokenize();
        expect(tokens.filter(t => t.type === TokenType.NUMBER).length).toBe(3);
    });

    it('throws on unterminated multi-line comment', () => {
        const lexer = new Lexer('let x = 1; /* unterminated comment');
        expect(() => lexer.tokenize()).not.toThrow(); // Should skip to EOF, not throw
    });

    it('skips comments inside code', () => {
        const lexer = new Lexer('let /* comment */ x = 1;');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toContain(TokenType.KEYWORD);
        expect(tokens.map(t => t.type)).toContain(TokenType.IDENTIFIER);
    });

    it('handles leading, trailing, and multiple whitespace', () => {
        const lexer = new Lexer('   let   x   =   1   ;   ');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toContain(TokenType.KEYWORD);
        expect(tokens.map(t => t.type)).toContain(TokenType.IDENTIFIER);
        expect(tokens.map(t => t.type)).toContain(TokenType.ASSIGNMENT);
        expect(tokens.map(t => t.type)).toContain(TokenType.NUMBER);
        expect(tokens.map(t => t.type)).toContain(TokenType.SEMICOLON);
    });

    it('handles very large input (stress test)', () => {
        const code = 'let x = 1; '.repeat(10000);
        const lexer = new Lexer(code);
        const tokens = lexer.tokenize();
        expect(tokens.length).toBeGreaterThan(10000);
    });
});

// --- Additional edge case tests ---
describe('Lexer - additional edge cases', () => {
    it('tokenizes unicode identifiers', () => {
        const lexer = new Lexer('const π = 3.14; let 变量 = 42;');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.value === 'π')).toBe(true);
        expect(tokens.some(t => t.value === '变量')).toBe(true);
    });

    it('tokenizes strings with escaped line breaks', () => {
        const lexer = new Lexer('"foo\\\nbar"');
        const tokens = lexer.tokenize();
        expect(tokens[0].type).toBe(TokenType.STRING);
    });

    it('tokenizes empty template literal', () => {
        const lexer = new Lexer('const t = ``;');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_STRING)).toBe(true);
    });

    it('tokenizes template literal with only expression', () => {
        const lexer = new Lexer('const t = `${foo}`;');
        const tokens = lexer.tokenize();
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_STRING)).toBe(true);
        expect(tokens.some(t => t.type === TokenType.TEMPLATE_EXPRESSION)).toBe(true);
    });

    it('tokenizes numbers with leading zeros', () => {
        const lexer = new Lexer('0123');
        const tokens = lexer.tokenize();
        expect(tokens[0].type).toBe(TokenType.NUMBER);
    });

    it('handles invalid numeric literals gracefully', () => {
        const lexer = new Lexer('0xGHI 0b102 0o89 1e 1e+ 1e-');
        // Should throw an error for invalid numbers
        expect(() => lexer.tokenize()).toThrow();
    });

    it('handles backslash at end of string', () => {
        const lexer = new Lexer('"foo\\\\"');
        const tokens = lexer.tokenize();
        expect(tokens[0].type).toBe(TokenType.STRING);
    });

    it('skips non-ASCII whitespace', () => {
        const lexer = new Lexer('\u00A0let\u2003x = 1;');
        const tokens = lexer.tokenize();
        expect(tokens.map(t => t.type)).toContain(TokenType.KEYWORD);
        expect(tokens.map(t => t.type)).toContain(TokenType.IDENTIFIER);
    });
}); 

File: legacy/backends/llvm/tests/lexer_test.cpp
#include "lexer/Lexer.h"
#include <gtest/gtest.h>
#include <string>
#include <vector>
#include <iostream>
#include <fstream>

using namespace superjs;

TEST(LexerTest, BasicTokens) {
    std::string source = "let x = 42;";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 6);  // 5 tokens + EOF
    EXPECT_EQ(tokens[0].kind, TokenKind::Let);
    EXPECT_EQ(tokens[1].kind, TokenKind::Identifier);
    EXPECT_EQ(tokens[2].kind, TokenKind::Equal);
    EXPECT_EQ(tokens[3].kind, TokenKind::Number);
    EXPECT_EQ(tokens[4].kind, TokenKind::Semicolon);
    EXPECT_EQ(tokens[5].kind, TokenKind::EndOfFile);
}

TEST(LexerTest, StringLiterals) {
    std::string source = "\"hello\" 'world'";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 3);  // 2 tokens + EOF
    EXPECT_EQ(tokens[0].kind, TokenKind::String);
    EXPECT_EQ(tokens[0].text, "\"hello\"");
    EXPECT_EQ(tokens[1].kind, TokenKind::String);
    EXPECT_EQ(tokens[1].text, "'world'");
    EXPECT_EQ(tokens[2].kind, TokenKind::EndOfFile);
}

TEST(LexerTest, Numbers) {
    std::string source = "42 3.14 0.5";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 4);  // 3 tokens + EOF
    EXPECT_EQ(tokens[0].kind, TokenKind::Number);
    EXPECT_EQ(tokens[0].text, "42");
    EXPECT_EQ(tokens[1].kind, TokenKind::Number);
    EXPECT_EQ(tokens[1].text, "3.14");
    EXPECT_EQ(tokens[2].kind, TokenKind::Number);
    EXPECT_EQ(tokens[2].text, "0.5");
    EXPECT_EQ(tokens[3].kind, TokenKind::EndOfFile);
}

TEST(LexerTest, Operators) {
    std::string source = "+ - * / % = == != < <= > >= && || !";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    // Debug output to file
    std::ofstream debug_file("lexer_debug.txt");
    debug_file << "\nActual tokens:" << std::endl;
    for (size_t i = 0; i < tokens.size(); ++i) {
        debug_file << "Token " << i << ": " << tokens[i].text << std::endl;
    }
    debug_file.close();
    
    ASSERT_EQ(tokens.size(), 16);  // 15 tokens + EOF
    const char* expected_texts[] = {
        "+", "-", "*", "/", "%", "=", "==", "!=", "<", "<=", ">", ">=", "&&", "||", "!"
    };
    for (int i = 0; i < 15; ++i) {
        EXPECT_EQ(tokens[i].text, expected_texts[i]);
    }
    EXPECT_EQ(tokens[15].kind, TokenKind::EndOfFile);
}

TEST(LexerTest, Comments) {
    std::string source = "// Single line comment\n/* Multi\nline\ncomment */";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 1);  // Just EOF
    EXPECT_EQ(tokens[0].kind, TokenKind::EndOfFile);
}

TEST(LexerTest, Keywords) {
    std::string source = "let const var function return if else while for";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 10);  // 9 tokens + EOF
    EXPECT_EQ(tokens[0].kind, TokenKind::Let);
    EXPECT_EQ(tokens[1].kind, TokenKind::Const);
    EXPECT_EQ(tokens[2].kind, TokenKind::Var);
    EXPECT_EQ(tokens[3].kind, TokenKind::Function);
    EXPECT_EQ(tokens[4].kind, TokenKind::Return);
    EXPECT_EQ(tokens[5].kind, TokenKind::If);
    EXPECT_EQ(tokens[6].kind, TokenKind::Else);
    EXPECT_EQ(tokens[7].kind, TokenKind::While);
    EXPECT_EQ(tokens[8].kind, TokenKind::For);
    EXPECT_EQ(tokens[9].kind, TokenKind::EndOfFile);
}

TEST(LexerTest, Identifiers) {
    std::string source = "x _y z123 _123";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 5);  // 4 tokens + EOF
    EXPECT_EQ(tokens[0].kind, TokenKind::Identifier);
    EXPECT_EQ(tokens[0].text, "x");
    EXPECT_EQ(tokens[1].kind, TokenKind::Identifier);
    EXPECT_EQ(tokens[1].text, "_y");
    EXPECT_EQ(tokens[2].kind, TokenKind::Identifier);
    EXPECT_EQ(tokens[2].text, "z123");
    EXPECT_EQ(tokens[3].kind, TokenKind::Identifier);
    EXPECT_EQ(tokens[3].text, "_123");
    EXPECT_EQ(tokens[4].kind, TokenKind::EndOfFile);
}

TEST(LexerTest, ErrorHandling) {
    std::string source = "\"unterminated string";
    Lexer lexer(source);
    auto tokens = lexer.tokenize();
    
    ASSERT_EQ(tokens.size(), 2);  // Error token + EOF
    EXPECT_EQ(tokens[0].kind, TokenKind::Error);
    EXPECT_EQ(tokens[0].text, "Unterminated string");
    EXPECT_EQ(tokens[1].kind, TokenKind::EndOfFile);
}

int main(int argc, char** argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
} 